modules/core/src/test/java/org/apache/ignite/internal/binary/streams/BinaryMemoryAllocatorTest.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.binary.streams;
+
+import org.apache.ignite.IgniteSystemProperties;
+import org.apache.ignite.testframework.junits.WithSystemProperty;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+import static org.apache.ignite.testframework.GridTestUtils.getFieldValue;
+
+/**
+ * BinaryMemoryAllocator self test.
+ */
+@WithSystemProperty(key = IgniteSystemProperties.IGNITE_MARSHAL_BUFFERS_RECHECK, value = "500")
+public class BinaryMemoryAllocatorTest extends GridCommonAbstractTest {
+    private static final int SMALL_OBJECT_SIZE = 1024;
+
+    private static final int LARGE_OBJECT_SIZE = 8 * 1024;
+
+    /**
+     * Ensure thread-local allocator shrinks cached buffer eventually.
+     */
+    @Test
+    public void threadLocalBufferCompaction() throws Exception {
+        resetThreadLocalBuffer();
+
+        BinaryMemoryAllocatorChunk buffer = threadLocalMemoryAllocatorBuffer();
+
+        // Allocate large object.
+        allocateThenRelease(buffer, LARGE_OBJECT_SIZE);
+
+        assertTrue(bufferSize(buffer) >= LARGE_OBJECT_SIZE);
```

Comment:
Looks like the test can be flaky.
1. ``assertTrue(bufferSize(buffer) >= LARGE_OBJECT_SIZE);`` can fail due to buffer will be shrinked earlier
2. `` Thread.sleep(1_000);`` don't bring any guarantee that shrink will be done.